cmds/fmap/lib/fmap.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fmap
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"strings"
+)
+
+var signature = []byte("__FMAP__")
+
+const (
+	FmapAreaStatic = 1 << iota
+	FmapAreaCompressed
+	FmapAreaReadOnly
+)
+
+type FMap struct {
+	FMapHeader
+	Areas []FMapArea
+}
+
+type FMapHeader struct {
+	Signature [8]uint8
+	VerMajor  uint8
+	VerMinor  uint8
+	Base      uint64
+	Size      uint32
+	Name      [32]uint8
+	NAreas    uint16
+}
+
+type FMapArea struct {
+	Offset uint32
+	Size   uint32
+	Name   [32]uint8
+	Flags  uint16
+}
+
+type FMapMetadata struct {
+	Start uint64
+}
+
+// FlagNames returns human readable representation of the flags.
+func FlagNames(flags uint16) string {
+	names := []string{}
+	m := []struct {
+		val  uint16
+		name string
+	}{
+		{FmapAreaStatic, "STATIC"},
+		{FmapAreaCompressed, "COMPRESSED"},
+		{FmapAreaReadOnly, "READ_ONLY"},
+	}
+	for _, v := range m {
+		if v.val&flags != 0 {
+			names = append(names, v.name)
+			flags -= v.val
+		}
+	}
+	// Write a hex value for unknown flags.
+	if flags != 0 || len(names) == 0 {
+		names = append(names, fmt.Sprintf("%#x", flags))
+	}
+	return strings.Join(names, "|")
+}
+
+func readField(r io.Reader, data interface{}) {
+	// The endianness might depend on your machine or it might not.
+	if err := binary.Read(r, binary.LittleEndian, data); err != nil {
+		log.Fatal("Unexpected EOF while parsing fmap")
+	}
+}
+
+// Read an FMap into the data structure.
+func ReadFMap(f io.Reader) (*FMap, *FMapMetadata) {
```

Comment:
You are completely right, I should return an error. That way, it is also possible to write unit tests for the error cases and get 100% coverage.

Secondly, I see a couple of advantages of not using a slice to store the flash and only using the `io.Reader`:

- `io.Reader` is an interface, so in theory, an implementation can read directly from the system's flash chip (through mmio or whatnot). Similarly, write with an `io.Writer`. This would make the fmap package a very powerful tool as it can verify and perform updates to the system's flash.
- A slice requires the entire flash to be present in memory. An `io.Reader` can buffer with O(1) memory usage.